### DISCUSSION TITLE
Unauthorized when login in via OIDC with Authentik

### DISCUSSION THREAD

**Discussion by crunchiersa** [photoprism/photoprism#4450] [Q&A]
Hi all,

I'm encountering a problem with my OIDC-Login-Config.
When I try to login via then configured OIDC after authenticating with Authentik (IdP - self hosted) I get returned to the login screen and an error is shown: invalid login (Ungültige Zugangsdaten - see attached screenshot).
![Screenshot 2024-08-06 at 19 32 32](https://github.com/user-attachments/assets/838eeaca-eac5-423a-a0c4-ebae330f0d64)

My OIDC-ENVs:
- PHOTOPRISM_SITE_URL=https://photo.my.domain/
- PHOTOPRISM_OIDC_URI=https://authentik.my.domain/application/o/photoprism/
- PHOTOPRISM_OIDC_CLIENT=${PHOTOPRISM_OIDC_CLIENT}
- PHOTOPRISM_OIDC_SECRET=${PHOTOPRISM_OIDC_SECRET}
- PHOTOPRISM_OIDC_PROVIDER=Authentik
- PHOTOPRISM_OIDC_REDIRECT=false
- PHOTOPRISM_OIDC_REGISTER=true
- PHOTOPRISM_OIDC_USERNAME=username
- PHOTOPRISM_OIDC_WEBDAV=true

redirect URL in Authentik is: https://photo.my.domain/api/v1/oidc/redirect
.well-known-config URI: https://authentik.my.domain/application/o/photoprism/.well-known/openid-configuration

docker exec -it photoprism photoprism show config gives the following:
```
|---------------|----------------------------------------------------------------------------------------------------------------------------------|
|     Name      |                                                              Value                                                               |
|---------------|----------------------------------------------------------------------------------------------------------------------------------|
| oidc-uri      | https://authentik.my.domain/application/o/photoprism/                                                                          |
| oidc-client   | {redacted}                                                                                         |
| oidc-secret   | ******************************************************************************************************************************** |
| oidc-scopes   | openid email profile address                                                                                                     |
| oidc-provider | Authentik                                                                                                                        |
| oidc-icon     | /static/img/oidc.svg                                                                                                             |
| oidc-redirect | false                                                                                                                            |
| oidc-register | true                                                                                                                             |
| oidc-username | preferred_username                                                                                                               |
| oidc-role     | guest                                                                                                                            |
| oidc-webdav   | true                                                                                                                             |
| disable-oidc  | false                                                                                                                            |
|---------------|----------------------------------------------------------------------------------------------------------------------------------|
```

docker exec -it photoprism photoprism users ls gives the following info:
```
|------------------|------------|-------|----------------|-------------|-----------|---------------------|----------|-------------|
|       UID        |  Username  | Role  | Authentication | Super Admin | Web Login |     Last Login      |  WebDAV  | Upload Path |
|------------------|------------|-------|----------------|-------------|-----------|---------------------|----------|-------------|
| us0rj7y1onwuzefj | photoprism | Admin | Local (2FA)    | Yes         | Enabled   | 2024-04-27 15:19:39 | Enabled  |             |
| usht07zixqflkfb0 | gmail      | Guest | OIDC           | No          | Enabled   |                     | Disabled |             |
|------------------|------------|-------|----------------|-------------|-----------|---------------------|----------|-------------|

```

The only relevant entry in the photoprism-log I see is this line: time="2024-08-06T17:19:53Z" level=debug msg="server: GET /api/v1/oidc/redirect?code=e2896c2c95e5470885775a97f3019d55&state=80273250-248f-4111-ade7-31e736314dc6 (401) [3.381745708s]"

The log-entry in Authentik shows:

Context
{
    "flow": "9611b91c76524146926b59b32ed5aada",
    "scopes": "email profile openid",
    "http_request": {
        "args": {
            "scope": "address email openid profile",
            "state": "eedc9c4c-76a2-4030-a76d-505a0b08d069",
            "client_id": "{redacted}",
            "redirect_uri": "https://photo.my.domain/api/v1/oidc/redirect",
            "response_type": "code",
            "code_challenge": "{redacted}",
            "code_challenge_method": "S256"
        },
        "path": "/api/v3/flows/executor/{redacted}/",
        "method": "GET",
        "request_id": "a95d59d4dbb240ffafb30f8be11b8f3f",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
    },
    "authorized_application": {
        "pk": "{redacted}",
        "app": "authentik_core",
        "name": "photoprism",
        "model_name": "application"
    }
}
User
{
    "pk": 16,
    "email": "{redacted}@gmail.com",
    "username": "gmail"
}

So it seems, that the OIDC is "recognised" as the User gets created and assigned in photoprism - but somehow photoprism doesn't permit the user to login? Did anybody get OIDC working with Authentik? Or maybe even solve this exact problem?

**Comment by lastzero** [comment]
Could it be that you try to log in with a username that already exists? Did you use the commands provided for checking login attempts yet?
- https://docs.photoprism.app/user-guide/users/cli/#viewing-login-attempts

**Reply by crunchiersa** [reply]
Thanks for responding! 
I assume you mean, that the account "gmail" already existed before configuring OIDC - if so then no, definitly not as the Account gmail was created after configuring OIDC by loggin in via authentik.
No, I didn't... definitly should have done that... Seems the problem exists due to the user being disabled... 
When I run docker exec photoprism photoprism audit logins gmail
The following is shown:
```
|-------------|----------|-------|-----------------------------------|------------|---------------------|
|  Client IP  | Username | Realm |              Status               | Last Login |      Failed At      |
|-------------|----------|-------|-----------------------------------|------------|---------------------|
| 10.10.30.52 | gmail    | oidc  | error: account disabled (2 times) |            | 2024-07-13 19:44:52 |
|-------------|----------|-------|-----------------------------------|------------|---------------------|
````
Is there a way to reenable (i.e. unlock?) the account? I couldn't find anything during a quick search on the CLI commands.

EDIT: I tried with a completely new test-user from Authentik - and it works... I reset the audit-logs and even tried giving the gmail-user a new password to login via user/password - but then login via OIDC still doesn't work... 

@lastzero I'm tempted to reset the user-table (via photoprism users reset --yes) and then login in only via OIDC and changing the roles via CLI. Would that work? I do understand that I would loose all current users.

**Comment by Tech-dea** [comment]
I have similar scenarios, but when I try to do photoprism user ls, I cannot see the OIDC user... but I can login (although there is nothing being seen), is that normal for OIDC users?

also when I use photoprism audit logins, I cannot see any OIDC user login... is that normal as well?

**Comment by WoPYUuNQck8Fm2ns** [comment]
It looks like PhotoPrism does not pick up the OIDC username configuration.

`PHOTOPRISM_OIDC_USERNAME=username` is provided as env var.

However, `photoprism show config` returns `| oidc-username | preferred_username `. I think it should show the env-configured `username` there instead of `preferred_username`.

I have the same problem when trying to use Synology SSO Server. It does not return a preferred_username field so I need to convince PhotoPrism to use the username instead.

In my case, this leads to `photoprism audit logins` returning:
 `| 192.168.176.2 |                                  | oidc  | error: username requ
ired to register |                     | 2024-10-21 00:11:34 | `

@lastzero do you think that's a bug?

**Reply by graciousgrey** [reply]
Possible values for the variable PHOTOPRISM_OIDC_USERNAME are: “preferred_username”, ‘name’, ‘nickname’ or ‘email’. Do none of these work with your Synology SSO server?

https://docs.photoprism.app/getting-started/config-options/

**Reply by lastzero** [reply]
@WoPYUuNQck8Fm2ns It doesn't look like "username" (instead of "preferred_username") is a standard OIDC claim, which is why it's not currently supported:

- https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

Note that it is not allowed to use a non-standard claim name, as this can lead to many conflicts and potential security issues, e.g. if the field is not unique.

**Reply by lastzero** [reply]
Our documentation has been updated to include a note on this:
- https://docs.photoprism.app/getting-started/advanced/openid-connect/#can-i-configure-a-custom-claim-for-the-preferred-username
- https://docs.photoprism.app/getting-started/faq/#can-i-configure-a-custom-claim-as-the-preferred-oidc-username

I hope this helps!

**Reply by WoPYUuNQck8Fm2ns** [reply]
@lastzero thanks for your replies.

From what I can tell, Synology SSO Server returns `email`, `iss`, `sub`, `username` in the access token and `email`, `sub`, `username` from the userinfo endpoint. I did not find a way to configure additional data points.

I changed the config to `PHOTOPRISM_OIDC_USERNAME=email` and see this value being used with `photoprism config show`. However, no luck, I still get `error: username required to register` in `photoprism audit logins`.


You mentioned that non-standard claim names are not allowed due to security and non-uniqueness issues.
None of these fields (`preferred_username`, `name`, `nickname` and `email`) are unique based on the spec.

I do not understand why you would restrict it to those. It blocks users with "non-standard" SSO servers and those who need to customize the mapping, e.g. by introducing a special photoprism_username claim name on the authz server.

A hint in the docs about making sure to choose a unique claim name should be enough and it is the admin's responsibility to choose a correct/unique claim name for the user mapping.

If there are reasons I don't know about for keeping the "allow list" of usable claim names, it would make sense to add `sub` to it. https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability says that this is the only unique data point (combined with `iss`). Therefore, it seems to be the only sensible claim name to use if you would like to prevent security issues due to non-unique values.

**Reply by lastzero** [reply]
Alright, I will remove the note on the uniqueness. What I meant was that much of the other standard fields are just not suitable for building a username. Also we implemented fallback logic which depends on the setting. Such logic doesn't exist for additional / custom claims and our client wouldn't know what type of information they represent either. Last but not least that is an initial implementation with a focus on security and getting it released within a limited amount of time. We therefore did not test or focus on any non standard features or claims your provider may support.

**Reply by lastzero** [reply]
Our documentation has been updated accordingly:

> You can choose between `preferred_username`, `name`, `nickname` and `email`, where `preferred_username` is the default. The other claims are used as fallback if no value is returned for the [configured claim](https://docs.photoprism.app/getting-started/advanced/openid-connect/#config-options).
> 
> Please note that it is currently not possible to use [other standard](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) or non-standard claims, as these may not be suitable for generating a username and no logic is implemented for doing so.

see https://docs.photoprism.app/getting-started/advanced/openid-connect/#can-i-configure-a-custom-claim-for-the-preferred-username

**Reply by WoPYUuNQck8Fm2ns** [reply]
Thank you for the clarification.

When I look at https://github.com/photoprism/photoprism/blob/develop/internal/auth/oidc/username.go#L34 it seems that `email_verified` claim is required in addition to the `email` claim. If `email_verified` is not provided by the SSO server (as in my case), it would explain why I still get the `error: username required to register` issue.

**Reply by lastzero** [reply]
Thanks for bringing this up! You are right that the email must be verified, so that we can (later) use it to send notifications and/or confirm the identity of users. If we don't insist on the verification then that's one of the things that could negativity impact security / trust.

That being said, I'm very surprised that none of these standard claims are set (except for an unverified email). Such a provider would probably not pass OIDC certification and should be improved / fixed in this regard. Is there any way you can report this to Synology or whoever maintains the software?

**Reply by lastzero** [reply]
Our documentation has been updated once again to include this information and a link to the related source code:
- https://docs.photoprism.app/getting-started/advanced/openid-connect/#what-if-my-provider-does-not-return-any-claims-for-generating-a-username
- https://docs.photoprism.app/developer-guide/api/oidc/#preferred-username
- https://github.com/photoprism/photoprism/blob/develop/internal/auth/oidc/username.go

**Reply by lastzero** [reply]
In a future release, we can consider removing the verification requirement if the email is only used for generating a username and not for sending notifications.

However, it must then be clear to admins and users that this is not the actual email address of the account. Alternatively, we could generate a completely random username for newly registered accounts if none of the standard claims are provided.

If there is a spec stating that 'username' is a common claim to be used by clients, we are happy to look at that as well.

**Comment by meister-eraser** [comment]
I have the same issue mentioed above but only for  LDAP User. Normaly created User in Authentik work fine.

      PHOTOPRISM_OIDC_URI: "https://authentik.zimmerei-loges.net/application/o/photoprism/"
      PHOTOPRISM_OIDC_CLIENT: "xxxxx
      PHOTOPRISM_OIDC_SECRET: "xxxx"
      PHOTOPRISM_OIDC_PROVIDER: "authentik"
      PHOTOPRISM_OIDC_REDIRECT: "true"
      PHOTOPRISM_OIDC_REGISTER: "true"
      PHOTOPRISM_OIDC_ICON: "/static/img/oidc.svg"
      PHOTOPRISM_DISABLE_TLS: "true"
      TZ: "Europe/Berlin"
     PHOTOPRISM_SITE_URL: "https://photoprism.zimmerei-loges.net"

The Strange thing is i don't see any Failed login for the users. Photoprism creates the user but it reports no failed login.

**Reply by meister-eraser** [reply]
Any ideas ?

**Comment by timo-reymann** [comment]
Encountering the same issues, just with a single user. Same symptoms: OIDC logs (Auth0) look fine, but Photoprism just returns a "Invalid credentials" after login with no logging or details: https://github.com/photoprism/photoprism/issues/5294
